data/trie/patriciaMerkleTrie.go

```diff
@@ -445,37 +449,71 @@ func (tr *patriciaMerkleTrie) ResetOldHashes() [][]byte {
 	return oldHashes
 }
 
-// Snapshot creates a new database in which the current state of the trie is saved.
-// If the maximum number of snapshots has been reached, the oldest snapshot is removed.
-func (tr *patriciaMerkleTrie) Snapshot() error {
+// Checkpoint adds the current state of the trie to the snapshot database
+func (tr *patriciaMerkleTrie) Checkpoint() error {
 	tr.mutOperation.Lock()
 	defer tr.mutOperation.Unlock()
 
+	if tr.root.isDirty() {
+		return ErrTrieNotCommited
+	}
+
 	if tr.snapshotInProgress {
 		return ErrSnapshotInProgress
 	}
 	tr.snapshotInProgress = true
 
-	err := tr.root.commit(false, 0, tr.db, tr.db)
+	if len(tr.snapshots) == 0 {
+		_, err := tr.newSnapshotDb()
+		if err != nil {
+			tr.snapshotInProgress = false
+			return err
+		}
+	}
+
+	lastSnapshotId := len(tr.snapshots) - 1
+	err := tr.root.commit(true, 0, tr.db, tr.snapshots[lastSnapshotId])
 	if err != nil {
+		tr.snapshotInProgress = false
 		return err
 	}
 
+	keys := tr.pruningBuffer
+	tr.pruningBuffer = make([][]byte, 0)
+	tr.snapshotInProgress = false
+
+	for i := range keys {
+		err = tr.removeFromDb(keys[i])
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Snapshot creates a new database in which the current state of the trie is saved.
+// If the maximum number of snapshots has been reached, the oldest snapshot is removed.
+func (tr *patriciaMerkleTrie) Snapshot() error {
+	tr.mutOperation.Lock()
+	defer tr.mutOperation.Unlock()
+
+	if tr.root.isDirty() {
+		return ErrTrieNotCommited
+	}
+
+	if tr.snapshotInProgress {
+		return ErrSnapshotInProgress
```

Comment:
Since all tr.snapshotInProgress calls are protected by tr.mutOperation these if branches that return ErrSnapshotInProgress are technically unreachable so we might drop tr.snapshotInProgress. The only use case remain in snapshot function L593. 
You should take into account that there is an edge case that might cause repetitive Snapshots triggers even though the previous snapshots are not done (caused by the L593 in conjunction with L596-L603 and L498-L554). Is this the intended behavior?

Reply:
There are cases where the if statements are reached, and used. For example, if I call `Snapshot()`, `tr.snapshotInProgress = true` will be set, but the mutex will be released at the end of this function, while  `tr.snapshotInProgress = false` will be set in the `go tr.snapshot(newTrie, db)` function call on line 555. If, before the setting of  `tr.snapshotInProgress = false` from the function that runs on a goroutine, we call `Snapshot()` again, we will acquire the released mutex, but when we reach the if from L506, the snapshotInProgress was not set to false, so we can not continue with the process of snapshoting.

Reply:
Technically, the snapshoting process has finished on L594. The pruning was buffered while snapshoting, but now (L594), thet the snapshot was finished, I can begin pruning from the bufferd values (L596-L603). If another snapshot is called before finishing the operations from L596-L603, the new snapshot will not interfere with these operations.